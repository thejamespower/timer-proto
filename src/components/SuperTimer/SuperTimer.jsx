import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Countdown from 'react-countdown-now'
import moment from 'moment'
import Button from '@material-ui/core/Button/Button'
import Card from '@material-ui/core/Card'
import CardActions from '@material-ui/core/CardActions'
import CardContent from '@material-ui/core/CardContent'
import Typography from '@material-ui/core/Typography'

class SuperTimer extends Component {
  static propTypes = {
    superTimer: PropTypes.shape({
      duration: PropTypes.string.isRequired,
      durationInSeconds: PropTypes.number.isRequired,
      active: PropTypes.bool.isRequired,
      complete: PropTypes.bool.isRequired,
      currentCount: PropTypes.number,
    }).isRequired,
    startSuperTimer: PropTypes.func.isRequired,
    tickSuperTimer: PropTypes.func.isRequired,
    completeSuperTimer: PropTypes.func.isRequired,
  }

  constructor(props) {
    super(props)

    this.handleSubmitClick = this.handleSubmitClick.bind(this)
    this.handleTick = this.handleTick.bind(this)
    this.handleComplete = this.handleComplete.bind(this)
  }

  shouldComponentUpdate(nextProps) {
    const { superTimer: { duration, active } } = this.props
    return nextProps.superTimer.duration !== duration || nextProps.superTimer.active !== active
  }

  handleSubmitClick() {
    const { startSuperTimer, superTimer: { duration } } = this.props
    if (duration === '00:00:00') return
    startSuperTimer()
  }

  handleTick({ total }) {
    const { tickSuperTimer } = this.props
    tickSuperTimer(total / 1000)
  }

  handleComplete() {
    const { completeSuperTimer } = this.props
    completeSuperTimer()
  }

  render() {
    const { superTimer } = this.props
    const {
      duration, active, complete,
    } = superTimer
    return (
      <Card>
        <CardContent>
          <Typography color="textSecondary" gutterBottom>
            Total: {duration}
          </Typography>
          {active && (
          <Countdown
            date={moment().add(moment.duration(duration)).toDate()}
            dayInHours
            onTick={this.handleTick}
            onComplete={this.handleComplete}
          />)}
          {complete && <p>All done!</p>}
        </CardContent>
        <CardActions>
          <Button
            disabled={duration === '00:00:00'}
            type="submit"
            variant="contained"
            color="primary"
            onClick={this.handleSubmitClick}
          >Start
          </Button>
        </CardActions>
      </Card>
    )
  }
}

export default SuperTimer
