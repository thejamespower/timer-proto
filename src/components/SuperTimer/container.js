import { connect } from 'react-redux';

import { bindActionCreators } from 'redux';
import SuperTimer from './SuperTimer';
import {
  startSuperTimer,
  tickSuperTimer,
  completeSuperTimer,
} from '../../action-creators';

const mapStateToProps = state => ({
  superTimer: state.timers.superTimer,
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      startSuperTimer,
      tickSuperTimer,
      completeSuperTimer,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(SuperTimer);
