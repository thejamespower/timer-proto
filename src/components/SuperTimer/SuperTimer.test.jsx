import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import Button from '@material-ui/core/Button/Button'
import SuperTimer from './SuperTimer'

describe('SuperTimer', () => {
  let props

  beforeEach(() => {
    props = {
      superTimer: {
        duration: '00:00:00',
        durationInSeconds: 0,
        active: false,
        currentCount: null,
        timeElapsed: 0,
        complete: false,
      },
      startSuperTimer: () => null,
      tickSuperTimer: () => null,
      completeSuperTimer: () => null,
    }
  })

  describe('render', () => {
    it('renders default state', () => {
      const component = renderer.create(
        <SuperTimer {...props} />,
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('renders duration state', () => {
      props.superTimer.duration = '00:00:01'

      const component = renderer.create(
        <SuperTimer {...props} />,
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('renders active state', () => {
      props.superTimer.active = true

      const component = renderer.create(
        <SuperTimer {...props} />,
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('renders complete state', () => {
      props.superTimer.complete = true

      const component = renderer.create(
        <SuperTimer {...props} />,
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('behaviour', () => {
    it('handles valid Button onClick', () => {
      props.startSuperTimer = jest.fn()
      props.superTimer.duration = '00:00:01'
      props.superTimer.durationInSeconds = 1

      const wrapper = shallow(<SuperTimer {...props} />)
      const button = wrapper.find(Button)

      button.prop('onClick')()

      expect(props.startSuperTimer).toHaveBeenCalled()
    })

    it('handles invalid Button onClick', () => {
      props.startSuperTimer = jest.fn()

      const wrapper = shallow(<SuperTimer {...props} />)
      const button = wrapper.find(Button)

      button.prop('onClick')()

      expect(props.startSuperTimer).not.toHaveBeenCalled()
    })
  })
})
