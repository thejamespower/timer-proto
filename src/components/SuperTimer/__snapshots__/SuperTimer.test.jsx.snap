// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuperTimer render renders active state 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
>
  <div
    className="MuiCardContent-root-29"
  >
    <p
      className="MuiTypography-root-30 MuiTypography-body1-39 MuiTypography-colorTextSecondary-63 MuiTypography-gutterBottom-57"
    >
      Total: 
      00:00:00
    </p>
    <span>
      0
      :
      00
      :
      00
      :
      00
    </span>
  </div>
  <div
    className="MuiCardActions-root-65"
  >
    <button
      className="MuiButtonBase-root-93 MuiButtonBase-disabled-94 MuiButton-root-67 MuiButton-contained-78 MuiButton-containedPrimary-79 MuiButton-raised-81 MuiButton-raisedPrimary-82 MuiButton-disabled-87 MuiCardActions-action-66"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="submit"
    >
      <span
        className="MuiButton-label-68"
      >
        Start
      </span>
    </button>
  </div>
</div>
`;

exports[`SuperTimer render renders complete state 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
>
  <div
    className="MuiCardContent-root-29"
  >
    <p
      className="MuiTypography-root-30 MuiTypography-body1-39 MuiTypography-colorTextSecondary-63 MuiTypography-gutterBottom-57"
    >
      Total: 
      00:00:00
    </p>
    <p>
      All done!
    </p>
  </div>
  <div
    className="MuiCardActions-root-65"
  >
    <button
      className="MuiButtonBase-root-93 MuiButtonBase-disabled-94 MuiButton-root-67 MuiButton-contained-78 MuiButton-containedPrimary-79 MuiButton-raised-81 MuiButton-raisedPrimary-82 MuiButton-disabled-87 MuiCardActions-action-66"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="submit"
    >
      <span
        className="MuiButton-label-68"
      >
        Start
      </span>
    </button>
  </div>
</div>
`;

exports[`SuperTimer render renders default state 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
>
  <div
    className="MuiCardContent-root-29"
  >
    <p
      className="MuiTypography-root-30 MuiTypography-body1-39 MuiTypography-colorTextSecondary-63 MuiTypography-gutterBottom-57"
    >
      Total: 
      00:00:00
    </p>
  </div>
  <div
    className="MuiCardActions-root-65"
  >
    <button
      className="MuiButtonBase-root-93 MuiButtonBase-disabled-94 MuiButton-root-67 MuiButton-contained-78 MuiButton-containedPrimary-79 MuiButton-raised-81 MuiButton-raisedPrimary-82 MuiButton-disabled-87 MuiCardActions-action-66"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="submit"
    >
      <span
        className="MuiButton-label-68"
      >
        Start
      </span>
    </button>
  </div>
</div>
`;

exports[`SuperTimer render renders duration state 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
>
  <div
    className="MuiCardContent-root-29"
  >
    <p
      className="MuiTypography-root-30 MuiTypography-body1-39 MuiTypography-colorTextSecondary-63 MuiTypography-gutterBottom-57"
    >
      Total: 
      00:00:01
    </p>
  </div>
  <div
    className="MuiCardActions-root-65"
  >
    <button
      className="MuiButtonBase-root-93 MuiButton-root-67 MuiButton-contained-78 MuiButton-containedPrimary-79 MuiButton-raised-81 MuiButton-raisedPrimary-82 MuiCardActions-action-66"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="submit"
    >
      <span
        className="MuiButton-label-68"
      >
        Start
      </span>
      <span
        className="MuiTouchRipple-root-96"
      />
    </button>
  </div>
</div>
`;
